### ISSUE TITLE
[ENHANCEMENT] features for Arabic and Romanian

### ISSUE THREAD

**Issue by AlexKKhalil** [danielgarthur/neanes#134]
First, this is an excellent program! It is more simple and intuitive than Melodos and much, much faster than typing everything out in MS Word with the various font sets floating around out there. I've been writing out music in English and Romanian. 

I'd note a couple of issues in Romanian that would be helpful: sometimes there are within-syllable hyphens
("alină-l Fecioară" for example). Currently, if I type a hyphen, it moves me to the next neume so I have been using "=" signs to get around that for now. It would be great to have a workaround built in (e.g. "press shift + hyphen to keep the hyphen in single syllable). Also, for some reason I can't type "á" and have had to resort to "ά" so that'd be great, also.

OK, so a bigger thing would be the ability to type in Arabic. There are two approaches: 
first, the most common one involves neumes read left-to-right as usual but each Arabic syllable reading right to left. Many chanters are used to this but it's a weird system and obscures the text, even for fluent speakers. Lay people can't even tell what hymn is written by looking at these weirdly broken up words. Still, due to ubiquity it would be really helpful to be able to do this.  

A less-common but more elegant approach is to reverse a few neumes, like isons, apostrophes, varia, omalon, etc, to enable typing the melos from right to left. This way, one could type in the lyrics with a normal Arabic font and the words would become really clear. There are a few examples of this floating around and I'd be happy to send some along if it'd be helpful. 
I know these are big requests but it would enable a lot of music writing to take place. 
Thanks!
Alex

**Comment by danielgarthur** [comment]
Hyphens and underscores can now be added to lyrics without causing a melisma in release `0.3.0-beta.20`. Hold <kbd>Ctrl</kbd> while pressing the hyphen (or if using Mac, hold <kbd>Alt</kbd>).

For `á`, could you describe how you enter this character? On an US International Keyboard, or on a Spanish keyboard, for example, I can type <kbd>'</kbd>, release, then press <kbd>a</kbd> to get `á`. But I'm not sure how to do it on a Romanian keyboard. You could also paste the character into the lyrics, at least until we figure out why it's not working the normal way.

For Arabic support, a few things would be helpful:

- Examples of reversed neumes. Is every neume reversed, or are some left the normal direction?
- Typed out syllables of a simple hymn, such as the Trisagion. I don't know Arabic so being able to paste in the syllables for testing would be helpful.
- What is a good font family and font size to test with? To me, Arabic letters always look tiny, so I don't have a good feel for what is considered a "normal, readable size" for a native Arabic reader.
- Does the app currently work for Arabic left-to-right, or is there work that needs to be done there, too? From what I can tell, I can switch to an Arabic keyboard and type right to left. But I don't know Arabic well enough to know if any thing might be missing or broken. See below for an example:
<img width="120" alt="image" src="https://github.com/danielgarthur/neanes/assets/4132779/206dc5f3-f1cb-4418-8025-c9875ef5df68">

- Are there any peculiarities that I should know about Arabic chant? For example, see the picture below. Doesn't the highlighted neume contain two syllables (`ilāh`)? Why are there two syllables allowed on a single neume? Or is my Arabic so bad that I'm simply misreading?
<img width="521" alt="image" src="https://github.com/danielgarthur/neanes/assets/4132779/0ba7f840-0696-409e-9606-1cfb6e69985e">

**Comment by AlexKKhalil** [comment]
Hi Daniel,

Thanks for this. I had missed the update about hyphens and underlines. For á (both for my Romanian and US keyboards), I have to press and hold “a” somewhat longer than usual and a menu will appear like this:
[A screenshot of a computer  Description automatically generated with low confidence]

Pressing ‘ and holding Alt will give me æ. When I try what I described above (with the menu) or copying and pasting an á from a text editor into Neanes, I get this: [A picture containing metalware, lever, typography  Description automatically generated]

Regarding Arabic: you’re right! I can pretty much type everything in using the melody-goes-left-to-right/syllables-go-right-to-left method. Here is an example (first line of 4th mode kekragaria from Mitri al Murr’s scores):
***@***.***

Was that possible in earlier versions? I had thought it wasn’t. The Melkites use Byzantine notation going right to left, which makes it a bit easier to parse the words (and avoids repeated letters, as in the example you provided). All it requires would be for a few compound neumes to be switched to the opposite sequence (those including kentemata + ison, elaphron, etc). Someone made a font set for this and Fr Ephrem posted it here awhile back:

Here's an example of what I mean:
[A picture containing text, font, line, white  Description automatically generated]

Thanks!


From: danielgarthur ***@***.***>
Date: Saturday, 17 June 2023 at 04:06
To: danielgarthur/neanes ***@***.***>
Cc: AlexKKhalil ***@***.***>, Author ***@***.***>
Subject: Re: [danielgarthur/neanes] [ENHANCEMENT] features for Arabic and Romanian (Issue #134)

Hyphens and underscores can now be added to lyrics without causing a melisma in release 0.3.0-beta.20. Hold Ctrl while pressing the hyphen (or if using Mac, hold Alt).

For á, could you describe how you enter this character? On an US International Keyboard, or on a Spanish keyboard, for example, I can type ', release, then press a to get á. But I'm not sure how to do it on a Romanian keyboard. You could also paste the character into the lyrics, at least until we figure out why it's not working the normal way.

For Arabic support, a few things would be helpful:

  *   Examples of reversed neumes. Is every neume reversed, or are some left the normal direction?
  *   Typed out syllables of a simple hymn, such as the Trisagion. I don't know Arabic so being able to paste in the syllables for testing would be helpful.
  *   What is a good font family and font size to test with? To me, Arabic letters always look tiny, so I don't have a good feel for what is considered a "normal, readable size" for a native Arabic reader.
  *   Does the app currently work for Arabic left-to-right, or is there work that needs to be done there, too? From what I can tell, I can switch to an Arabic keyboard and type right to left. But I don't know Arabic well enough to know if any thing might be missing or broken. See below for an example:
[image]<https://user-images.githubusercontent.com/4132779/246577292-206dc5f3-f1cb-4418-8025-c9875ef5df68.png>

  *   Are there any peculiarities that I should know about Arabic chant? For example, see the picture below. Doesn't the highlighted neume contain two syllables (ilāh)? Why are there two syllables allowed on a single neume? Or is my Arabic so bad that I'm simply misreading?
[image]<https://user-images.githubusercontent.com/4132779/246577146-0ba7f840-0696-409e-9606-1cfb6e69985e.png>

**Comment by danielgarthur** [comment]
I think I know what the issue is. The default font for lyrics is the `Omega` font that came with Fr. Ephraim's original font package. Unfortunately that font is missing some accented characters. Try going into Page Setup and changing the lyrics font to something common, like Arial or Helvetica.

**Comment by basil** [comment]
IIRC the reason for the existence of EZ Omega was to add half-width underscores for drawing word extensions. If Neanes is using a method other than underscores to draw word extensions, EZ Omega may have outlived its usefulness. CG Omega (which EZ Omega was based on) is itself a cheap derivative of [Optima](https://en.wikipedia.org/wiki/Optima).

**Comment by basil** [comment]
I suggest the title of this ticket be changed to "Right-to-left support for neumes" (aka Melkite mode), since AFAICT the software works correctly in both Romanian and Arabic for left-to-right neumes and any Unicode lyrics.

The Melkite convention itself could possibly use some improvement I think. I don't know how firmly established it is, since I've never seen handwritten scores in the Melkite convention. The typeset scores I have seen that use the Melkite convention all use the font mentioned above.

Among other things, it seems strange to me that the vareia would not be inverted in the Melkite convention. It is visually designed to enclose a neume, so when it is not inverted but placed to the right of the neume it makes little sense to enclose white space.

**Comment by AlexKKhalil** [comment]
I agree with this. It was just the issue with the fonts that had given the impression that improvements were needed in those other areas. I agree with what you say, Basil, about the Melkite convention—and have seen numerous variations (in which different symbols were reversed)—but it’s an area that I think would be valuable to experiment with. Might be more trouble than it’s worth to set up, though. I don’t know how difficult it would be to create a right-to-left typing mode and probably few would use it.
Thanks!

From: Basil Crow ***@***.***>
Date: Sunday, 16 July 2023 at 03:40
To: danielgarthur/neanes ***@***.***>
Cc: AlexKKhalil ***@***.***>, Author ***@***.***>
Subject: Re: [danielgarthur/neanes] [ENHANCEMENT] features for Arabic and Romanian (Issue #134)

I suggest the title of this ticket be changed to "Right-to-left support for neumes" (aka Melkite mode), since AFAICT the software works correctly in both Romanian and Arabic for left-to-right neumes and any Unicode lyrics.

The Melkite convention itself could possibly use some improvement I think. I don't know how firmly established it is, since I've never seen handwritten scores in the Melkite convention. The typeset scores I have seen that use the Melkite convention all use the font mentioned above.

Among other things, it seems strange to me that the vareia would not be inverted in the Melkite convention. It is visually designed to enclose a neume, so when it is not inverted but placed to the right of the neume it makes little sense to enclose white space.

**Comment by mattsmi** [comment]
Gents, I'm the guy who wrote the Right-to-Left font for the Byzantine Music Project, which @AlexKKhalil mentions.
I'd love the chance to use Neanes, if it were able to support Arabic lyrics right to left, as well as the Byzantine neumes right to left (using my font mentioned above).
The best font for the lyric text in Arabic would be Noto Naskh Arabic from Google fonts. A distant second contender would be Simplified Arabic (from Microsoft, but licensed for general installation). The Microsoft font does not support all the common combinations, though, of base letters and diacritics in Arabic.
The issue you mentioned above with الله taking more than one syllable is because you were looking a music that is written from left to right and thus the Arabic had been broken up into syllables, which flow from left to right across the line, but each syllable must be read from right to left. This makes reading music terribly difficult.
It is much easier to have the music flow from right to left, which naturally supports the Arabic text.
Am happy to help in *any* way I can. Attached a sample page of the Kyrie Ekekraxa from the feast of the Dormition (Annunciation) celebrated on the eve of 15 August.

[Dormition_kyrie_ekekraxa_ar.pdf](https://github.com/danielgarthur/neanes/files/12343111/Dormition_kyrie_ekekraxa_ar.pdf)

**Comment by basil** [comment]
Why would you not invert the vareia in scores like the above? It is visually designed to enclose a neume, so when it is not inverted but placed to the right of the neume it makes little sense to enclose white space.

**Comment by mattsmi** [comment]
I suppose because we also chant in Greek and in English, and thus use all the symbols in their normal shapes.

When chanting in Arabic, we only change the compound symbols so that the actions they indicate flow correctly from right to left, as does the Arabic text.

If we were to begin changing everything, then note vareia, but should not the ison and clasma be flipped as well? And others? We thought that the change would be too great and would disturb the reading of the music while chanting. Thus the decision was mde only to change the compound symbols.

**Comment by mattsmi** [comment]
Tatweel is a standard practice for anything, where words must fit some other pattern, e.g. alignment under boxes, tables, etc.
Arabic cannot and should not be broken into syllables. The written form does not allow it and you do not see that anywhere.
That is another reason why the left-to-right syllabic Arabic forced to fit left-to-right Byzantine psaltiki is so difficult to read. The letters take on different forms when they are at the end of a word, rather than in the beginning.
By the way in Arabic, the lengthening of the stroke is more often called taTwiil ([tatˁwi:l] تَطْوِيل ).

**Comment by basil** [comment]
> I suppose because we also chant in Greek and in English

Just as it is inappropriate to retain melodic lines from the original Greek when this violates the accentuation of another language, so also it is inappropriate to retain Greek notational forms when they violate the typographic consistency of the score in a new context.

> we only change the compound symbols so that the actions they indicate flow correctly from right to left

But this violates the typographic consistency of the score.

> If we were to begin changing everything

Not everything needs to be changed. Things only need to be changed if the underlying reasoning behind the original typographic form is no longer valid.

> should not the ison and clasma be flipped as well?

I do not know without further study. The reasoning behind the original typographic form should be determined. If that reasoning is no longer valid, then a change should be made.

> And others?

Certainly. For example, the slope of the serifs and kentimata is left to right from top to bottom, while the slope of the Arabic font in your example is right to left from top to bottom. At minimum this implies a corresponding change to the serifs and kentimata of the music font.

> We thought that the change would be too great

On the contrary, the change in this convention is too small and lacks typographic consistency.

> Thus the decision was mde [sic] only to change the compound symbols

This decision seems premature. This convention needs further revision to implement typographic consistency. In its current form, it is not appropriate for a typographically advanced software project like Neanes.

**Comment by mattsmi** [comment]
I am sorry. I seems I may have not expressed myself well and perhaps incompletely.

The base neumes (e.g. ison, Elaphron) and the combining neumes (e.g. gorgons, eteron, clasma) were never changed from the way they appear when supporting Greek lyrics. Their shapes are the same regardless of the language supported and the direction of reading.

The only changes made were made to the compound neumes where the oligon supports two base neumes: since the order in which those supported base neumes is chanted is different, when reading from right to left, that order was changed.

Thus to the monks, they were not altering the basic neumes of Byzantine music, which were considered sacred.

This is not a new innovation. 

The example page I gave you, is from one of the psaltic books from the patriarchal protopsaltic and liturgical committee. All our official music, whether recently published or not, appears this way.

Thus the request to support it in your software, we hoped would bring little more effort than your supporting right-to-left rather than left-to-right writing of the neumes and text.

Please let me know, if there is anything else I can help with.

**Comment by mattsmi** [comment]
Since this method of writing Byzantine music with Arabic text has been in place for a couple of hundred years, it is unlikely that it will change. 

It did not bother or surprise Father Ephraim of St Anthony's monastery, when I was working with him in creating the right-to-left font about fifteen years ago. (I assume that your reference to Rule 84 refer to those rules collated by Father Ephraim of St Anthony's monastery?)

Will it be possible to support writing Byzantine music from right to left, supporting Arabic text in Neanes?

Many thanks for your time in analysing the issues, @basil .

**Comment by basil** [comment]
> Since this method of writing Byzantine music with Arabic text has been in place for a couple of hundred years, it is unlikely that it will change.

But humanity's understanding of regular tetrahedra changed after 1800 years, so time is not necessarily a supporting argument.

> It did not bother or surprise Father Ephraim of St Anthony's monastery, when I was working with him in creating the right-to-left font about fifteen years ago.

This does not imply the convention is linguistically, typographically, and orthographically consistent. It is not, for the reasons given earlier. In its current form, this convention and the scores produced with it are in need of revision.

**Comment by basil** [comment]
> Will it be possible to support writing Byzantine music from right to left, supporting Arabic text in Neanes?

Software support for this convention would appear to be premature as long as the issues described above remain unaddressed. Since technology is an accelerating force, we should apply discernment regarding its scope. The score you provided above contains numerous formulaic and orthographical errors.

> Many thanks for your time in analysing the issues, @basil .

Many thanks for your time as well @mattsmi.

**Comment by AlexKKhalil** [comment]
I would agree with Basil that, actually, there is some variability in how this is applied and it has yet to be done very well. I seem to recall having seen hand-written scores that have quite a few more neumes reversed. A key aspect of the typography would be to help one grasp the shape of a melody and recognize formulaic patterns at a glance. Interpreting right-to-left scores in Arabic in which only a handful of neumes are reversed can be confusing because the direction and shape of individual neumes, and combinations thereof, at a glance still reflect patterns that go from left to right. If more neumes were reversed (e.g. isons, slope of kentimata, apostrophes, etc) it might be possible to recognise more readily the familiar melodic structures of Byzantine music moving from right to left.

I would question the idea that the request is premature in that just as with English and many other languages, experimentation needs to happen. No one is instantly going to invent and dictate the perfect system. It needs to develop. If it were not difficult to achieve here (and this I do not know), this could become a flexible space to experiment. It would involve

a) the option to type a melody from right to left and,
b) a set of reversed neumes.

A user could place any neume in a right-to-left melody, reversed or not, and thus experiment with different configurations (anything from only compound neumes that would be out of order not reversed to almost everything, isons and all, reversed—reversed martyria would probably be the only step too far!). My motivation in the original query was more about experimentation. Technology doesn’t only need to be something that accelerates and ossifies certain practices (although there are examples in Byzantine chant of exactly this): it can also be a space to experiment. At the same time, this might be much too big of an ask for something that is both untried and quixotic (considering that most people, despite whatever issues arise from notational choices, appear to be fully entrenched in their own approach).

Best,

Alex


From: Basil Crow ***@***.***>
Date: Thursday, 17 August 2023 at 4:10 p.m.
To: danielgarthur/neanes ***@***.***>
Cc: AlexKKhalil ***@***.***>, Mention ***@***.***>
Subject: Re: [danielgarthur/neanes] [ENHANCEMENT] RTL support for Melkite scores (Issue #134)

Will it be possible to support writing Byzantine music from right to left, supporting Arabic text in Neanes?

Software support for this convention would appear to be premature as long as the issues described above remain unaddressed. Since technology is an accelerating force, we should apply discernment regarding its scope. The score you provided above contains numerous formulaic and orthographical errors.

Many thanks for your time in analysing the issues, @basil<https://github.com/basil> .

Many thanks for your time as well @mattsmi<https://github.com/mattsmi>.

**Comment by basil** [comment]
Yes experimentation needs to happen, but the problem with experimentation in software with a stable on-disk file format is that users who save data expect backwards compatibility in future versions. This is particularly true for open-source software, which tends to live a long time. Thus once support is added for a particular feature and a score using that feature is saved to disk, the feature stops being experimental in some sense and becomes part of the interface that the software must permanently support.

Over time, these vestigial experiments accrue and create nearly insurmountable maintenance problems. Another software project I work on is 18 years old, and many otherwise trivial projects explode in scope due to backwards compatibility constraints. And my experience is that no matter how many warnings and disclaimers you put on experimental software features, once a user starts relying on them that user will complain if you try to remove the feature later when it becomes hard to maintain. So I would recommend that any experimentation be done in a separate branch or fork rather than the main software deliverable.

**Comment by basil** [comment]
> One cannot simply reverse the direction of a writing system without significant changes to its graphemes.

FTR Leonardo da Vinci [reversed the direction of Latin script](https://www.mos.org/leonardo/activities/mirror-writing), making significant changes to its graphemes in the process.

**Comment by basil** [comment]
See also this page on [boustrophedonics](https://en.wikipedia.org/wiki/Boustrophedon). Of note is that some boustrophedon scripts employ mirroring, while others (including the constructed language Ithkuil by linguistic genius John Quijada) [retain their normal lateral orientation and are not laterally reversed](http://www.ithkuil.net/11_script.htm). My sense is that Byzantine notation likely shares more properties with the former than with the latter, but this needs to be investigated further by someone with linguistic training.

**Comment by basil** [comment]
The boustrophedon music notation envisioned by Jean-Jacques Rousseau:

![rousseau](https://github.com/danielgarthur/neanes/assets/29850/51d13623-1803-4df5-91ac-f43209def290)

The numerical system is not unlike [the Byzantine music notation reform proposed by George of Lesvos](https://anemi.lib.uoc.gr/metadata/a/c/7/metadata-247-0000107.tkl). Interestingly enough, Rousseau does mirror the Latin script for the lyrics in the same manner as da Vinci. The fact that he juxtaposes these conventions on the same line definitely indicates there is something going on here linguistically, as there must be some criteria for when to mirror and when not to mirror.

> I think it would make sense for Byzantine script to be mirrored when going right-to-left, considering that the contours of the neumes are largely derived from the movement of the hand.

I agree with this point at a high level as well as your previous point about recognizing formulaic patterns at a glance. But like you said, some experimentation is needed to examine every particular case and see how well it works in practice.

> As a counter example, classical Chinese can be written in either direction (and also top-to-bottom) without mirroring but that’s partly because each character represents generally an entire word: stroke direction and movement are very important but only within each character.

Same with Ithkuil.

> If, despite this, there is interest in developing a fork, then it would probably be best to adopt/enable whatever system is being used in the Melkite church today (despite noted flaws) with the option to mirror more neumes and see what happens.

Just as I think it would be unlikely for someone to start with e.g. Kazan's "Lord I have cried" and finish with e.g. Papa Ephraim's version, I think it would be unlikely for someone to start with the existing Melkite convention and finish with a linguistically, typographically, and orthographically consistent convention. Seems more straightforward to start with a clean slate. I think it would be highly undesirable to promote a convention that results in the flawed score shown above. If the example didn't contain so many formulaic and orthographic errors I might have felt less strongly about this.

**Comment by danielgarthur** [comment]
I'm converting this into a discussion. Feel free to continue working this out. My main concern is that, assuming all the details are worked for the convention, could this be maintained as a single project or ought it to be split into a separate project and how do we keep them in sync without too much effort.
